### PR TITLE
Generate Docs API

### PR DESCRIPTION
**Pull request by ScottCarda-MS**
Adds a new module to the qsc for docs generation and moves the display module from the language server to this new module. The docs generation is used by the npm to build library API docs from the doc comments of our Q# library code and puts those docs in the npm package. This is done when the npm package is built.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,7 +11,7 @@
   },
   "scripts": {
     "eslint:check": "eslint -c ./.eslintrc.cjs --max-warnings 0 ./",
-    "prettier:check": "prettier -c --no-config ./",
+    "prettier:check": "prettier -c --no-config --ignore-path ./",
```
Comment (billti) [inline package.json:14]: Is this change needed? It was picking up the prettierignore file before file OK before. Was something broken?

File: language_service/src/generate_docs/tests.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+use crate::compilation::Compilation;
+
+#[test]
+fn callable_unit_types() {
+    let compilation = Compilation::new(
+        &[],
+        qsc::PackageType::Exe,
+        qsc::target::Profile::Unrestricted,
+    );
+    generate_docs(&compilation);
```
Comment (billti) [inline language_service/src/generate_docs/tests.rs:16]: Does this verify anything about the output? Or just that the call succeeded?

File: language_service/src/generate_docs.rs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+use std::rc::Rc;
+
+// Warning: this path gets deleted on each run. Use carefully!
+const GENERATED_DOCS_PATH: &str = "../npm/generated_docs";
+
+pub(crate) fn generate_docs(compilation: &Compilation) {
+    let display = &CodeDisplay { compilation };
+
+    delete_existing_docs();
+    fs::create_dir(GENERATED_DOCS_PATH).expect("Unable to create directory for generated docs");
```
Comment (billti) [inline language_service/src/generate_docs.rs:23]: It's a little odd to be doing file system operations in the language service, being this is usually loaded into a WebWorker in a browser context. I'd expect it to return the content for some other host to do the writing for.

What calls into this? i.e., What is the mechanism by which this is invoked to generate the docs?

Seems maybe we should expose a language service API to get the structure & content of all docs, and then have a Node.js invocation to call into this to fetch the content and write it to disk.

File: package.json (hunk 1 above)
Comment (ScottCarda-MS) [reply package.json]: Nope I meant to remove that. Thanks!

File: language_service/src/generate_docs/tests.rs (hunk 2 above)
Comment (ScottCarda-MS) [reply language_service/src/generate_docs/tests.rs:16]: This isn't a real unit test. This is the only way, currently, I have of triggering the generator. I'll need to have more clear direction on how we intend to use this and what to hook it up to before merging in.

File: npm/generated_docs/Microsoft.Quantum.Arrays/All.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
+---
```
[the remark this answers is not in the document]
Comment (billti) [reply npm/generated_docs/Microsoft.Quantum.Arrays/All.md:1]: Yeah, good call out. I think it's fine to leave in for the PR review so we can see what's been generated, but it should be excluded from the checked in code.

Also, to be in the npm package, you'll want to add the docs folder to the 'files' object in the package.json from the npm package.

File: language_service/src/lib.rs [hunk 5]
```diff
@@ -7,7 +7,6 @@
 mod compilation;
 pub mod completion;
 pub mod definition;
-mod display;
 pub mod hover;
```
Comment (billti) [inline language_service/src/lib.rs:10]: Does anything use this module/file now in this crate? I don't see the file being deleted.

File: language_service/Cargo.toml [hunk 6]
```diff
@@ -24,6 +24,7 @@ enum-iterator = { workspace = true }
 regex-lite = { workspace = true }
 qsc_project = { path = "../compiler/qsc_project", features = ["async"] }
 async-trait = { workspace = true }
+chrono = { workspace = true }
```
Comment (billti) [inline language_service/Cargo.toml:27]: Do you still need chrono in this crate?

File: compiler/qsc_doc_gen/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+qsc_data_structures = { path = "../qsc_data_structures" }
+qsc_eval = { path = "../qsc_eval" }
+qsc_frontend = { path = "../qsc_frontend" }
+qsc_ast = { path = "../qsc_ast" }
+qsc_fir = { path = "../qsc_fir" }
+qsc_hir = { path = "../qsc_hir" }
+qsc_passes = { path = "../qsc_passes" }
+qsc_project = { path = "../qsc_project", features = ["fs"] }
+rustc-hash = { workspace = true }
+thiserror = { workspace = true }
```
Comment (billti) [inline compiler/qsc_doc_gen/Cargo.toml:36]: Does this crate really depend on all of the above?

File: language_service/src/generate_docs.rs (hunk 3 above)
Comment (ScottCarda-MS) [reply language_service/src/generate_docs.rs]: File I/O has been moved to the npm layer, per your commits.

File: npm/generated_docs/Microsoft.Quantum.Arrays/All.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply npm/generated_docs/Microsoft.Quantum.Arrays/All.md]: The docs have been excluded from the PR, per your suggestion.

File: language_service/src/lib.rs (hunk 5 above)
Comment (ScottCarda-MS) [reply language_service/src/lib.rs:10]: Not sure if this is an outdated comment. The display module is removed from the language server.

File: language_service/Cargo.toml (hunk 6 above)
Comment (ScottCarda-MS) [reply language_service/Cargo.toml]: I've cleaned up the Cargo.toml files for the language server and the new docs generation module.

File: compiler/qsc_doc_gen/Cargo.toml (hunk 7 above)
Comment (ScottCarda-MS) [reply compiler/qsc_doc_gen/Cargo.toml]: I've cleaned up the Cargo.toml files for the language server and the new docs generation module.

File: compiler/qsc_doc_gen/src/generate_docs.rs [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+    })
+}
+
+fn generate_toc(
+    map: &FxHashMap<Rc<str>, Vec<String>>,
+    file_map: &mut FxHashMap<Arc<str>, Arc<str>>,
+) {
+    let header = "
+# This file is automatically generated.
+# Please do not modify this file manually, or your changes may be lost when
```
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply compiler/qsc_doc_gen/src/generate_docs.rs:304]: This was copied over from the existing docs, but you are right, it should say "will be lost".

File: wasm/src/lib.rs [hunk 9]
```diff
@@ -324,6 +325,28 @@ pub fn check_exercise_solution(
...
+    filename: String,
+    contents: String,
+}
+
+#[wasm_bindgen]
+pub fn generate_docs() -> JsValue {
+    let docs = qsc_doc_gen::generate_docs::generate_docs();
+    let mut result: Vec<DocFile> = vec![];
+
+    // Create a JsObject with key value pairs
```
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply wasm/src/lib.rs]: removed

File: compiler/qsc_doc_gen/src/generate_docs.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply compiler/qsc_doc_gen/src/generate_docs.rs]: updated

File: compiler/qsc_doc_gen/src/generate_docs.rs [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+        topic: "managed-reference".to_string(),
+        kind,
+        namespace: ns,
+        name,
+        summary: parse_doc_for_summary(&item.doc),
+        signature,
+    })
+}
+
+fn generate_toc(
```
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply compiler/qsc_doc_gen/src/generate_docs.rs:299]: Table of Contents. I added a small doc comment to explain this in code.

File: compiler/qsc_doc_gen/src/display.rs [hunk 11]
```diff
@@ -1,63 +1,79 @@
...
 };
 use regex_lite::Regex;
 use std::{
     fmt::{Display, Formatter, Result},
     rc::Rc,
 };
 
-pub(crate) struct CodeDisplay<'a> {
-    pub(crate) compilation: &'a Compilation,
+pub trait Lookup {
```
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply compiler/qsc_doc_gen/src/display.rs:17]: added doc comments

File: Cargo.toml [hunk 12]
```diff
@@ -31,6 +32,7 @@ version = "0.0.0"
 
 [workspace.dependencies]
 bitflags = "2.4.2"
+chrono = "0.4.34"
```
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply Cargo.toml]: I've limited the dependencies for chrono as much as I can before errors arise. I ended up needing `alloc`, `clock`, and `wasmbind`.

File: npm/package.json [hunk 13]
```diff
@@ -23,15 +23,17 @@
     "./ux": "./ux/index.ts"
   },
   "scripts": {
-    "build": "npm run generate && npm run build:tsc",
+    "build": "npm run docs && npm run generate && npm run build:tsc",
```
[the remark this answers is not in the document]
Comment (billti) [reply npm/package.json:26]: That may be a more general question we can tackle outside of this, but the generate script writes to disk synchronously also. It still runs pretty quick in my unscientific experience.

File: npm/generate_docs.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// @ts-check
+
+import { existsSync, mkdirSync, writeFileSync } from "node:fs";
```
[the remark this answers is not in the document]
Comment (billti) [reply npm/generate_docs.js:6]: See prior discussion. At some point we may want to make greater use of this API for an in-editor 'cargo docs' like feature. In editor (and in Wasm in general) you can't write to the file system. I don't see building a new executable just for generating docs as being any less complex than 56 lines of JavaScript.

File: Cargo.toml (hunk 12 above)
[the remark this answers is not in the document]
Comment (billti) [reply Cargo.toml]: I was debating if we should just remove this entirely in the Rust side to avoid this dependency, and have the the script that writes to disk add the timestamp to the files (which is just a line or two in the Node script).

File: npm/generate_docs.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (minestarks) [reply npm/generate_docs.js:6]: > I don't see building a new executable just for generating docs as being any less complex than 56 lines of JavaScript.

While I'm 100% on board with the wasm approach, I feel like you're overstating the complexity of adding a binary in Rust. It's like 10 lines of code - you add a main() method and a couple of lines in cargo.toml.

I do understand we need the WASM API for different reasons though, not because it's less complex.

**Review by minestarks — APPROVED** [review]
Signing off on the language service-adjacent parts, didn't get a chance to look at the actual doc generation utility.

File: npm/generate_docs.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (billti) [reply npm/generate_docs.js:6]: I agree. I'm not overstating it. I get that's trivial _also_. Just trying to make the point this is trivial, and not "unnecessary complexity" per the original comment.

File: .prettierignore [hunk 15]
```diff
@@ -14,6 +14,7 @@ __pycache__/
 /katas/**/*.md
 /library/
 /npm/dist/
+/npm/generated_docs/
```
[the remark this answers is not in the document]
Comment (billti) [reply .prettierignore:17]: This should be '/npm/docs' now. It doesn't go into 'generated_docs' any more.

File: language_service/src/generate_docs/tests.rs (hunk 2 above)
Comment (ScottCarda-MS) [reply language_service/src/generate_docs/tests.rs]: This has been replaced with an actual unit test now that we have the logic hooked up to npm.

File: Cargo.toml (hunk 12 above)
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply Cargo.toml]: Good suggestion for moving the timestamping to the npm package Bill. I've gone ahead and made that change and dropped the chrono dependencies altogether.

File: .prettierignore (hunk 15 above)
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply .prettierignore]: The docs are already ignored by git, but I am now noticing that the path was changed from `/npm/generated_docs/` to `/npm/docs/`, which tells me that this line is not actually needed anymore.

File: npm/generate_docs.js [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+if (!docs || !docs.length) {
+  throw new Error("No docs generated");
+}
+
+var today = new Date();
+var dd = String(today.getDate()).padStart(2, "0");
+var mm = String(today.getMonth() + 1).padStart(2, "0"); //January is 0!
+var yyyy = today.getFullYear();
+var today_str = mm + "/" + dd + "/" + yyyy + "12:00:00 AM";
```
Comment (billti) [inline npm/generate_docs.js:33]: So there is no separator between yyyy and "12:00:00"? Does this look like "03/01/202412:00:00 AM" then? Doesn't seem right.

Comment (ScottCarda-MS) [reply npm/generate_docs.js]: Yep you are right, there should be a space. Fixed.

**Review by billti — APPROVED** [review]

**Review by swernli — APPROVED** [review]
